lib/Frontend/CompilerInvocation.cpp

```diff
@@ -1701,7 +1701,25 @@ static void ParseLangArgs(LangOptions &Opts, ArgList &Args, InputKind IK,
   Opts.PascalStrings = Args.hasArg(OPT_fpascal_strings);
   Opts.VtorDispMode = getLastArgIntValue(Args, OPT_vtordisp_mode_EQ, 1, Diags);
   Opts.Borland = Args.hasArg(OPT_fborland_extensions);
-  Opts.CheckedC = Args.hasArg(OPT_fcheckedc_extension);
+  if (Args.hasArg(OPT_fcheckedc_extension)) {
+    const char *disallowed = nullptr;
+    if (Opts.CUDA)
+      disallowed = "CUDA";
+    else if (Opts.OpenCL)
+      disallowed = "OpenCL";
+    else if (Opts.CPlusPlus)
+      disallowed = "C++";
+    else if (Opts.ObjC1 || Opts.ObjC2)
+      disallowed = "Objective C";
```

Comment:
These options are mutually exclusive, except that Objective C code may be compiled side-by-side with C++ code.  CUDA, OpenCL, and Objective C are all treated as different languages by clang (see clang\include\clang\Basic\LangOptions.def for a list of the language options).   I've updated the message for Objective C when compiled side-by-side with C++.  I also switched to using std::string instead of raw C strings, just to clean up the code.